feat(cli): CcApi context provider can be configured to fail if listing does not find a specific count of resources

Tietew opened the pull request:
Fixes #257

As described in the issue, `CcApiContextProviderPlugin` should have the way to ensure the result has exact one resource.

This PR adds the `expectedMatchCount` option to restrict the length of results in `listResources()`.

Unit tests are added and modified ensure the changes.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license

Tietew committed:
ensure exact one result in listResources

Tietew committed:
throw ConttextProviderError

mrgrain reviewed (commented):
@rix0rrr this looks okay to me, but you were more involved with the original version of this

rix0rrr commented:
I agree with you that persisting a 0-length array may make it harder to recover from problems in the future.

But I don't agree that a >=1 length array is automatically a failure. It is up the `lookup` method to reject >=1 element if it is expecting exactly one. It is not up to the provider to say "I will never support looking up more than one resource". For example, securitygroups associated with another resource are a perfectly legit resource to look up.

rix0rrr reviewed (changes requested):
I agree with you that persisting a 0-length array may make it harder to recover from problems in the future. 

But I don't agree that a >=1 length array is automatically a failure. It is up the `lookup` method to reject >=1 element if it is expecting exactly one. It is not up to the provider to say "I will never support looking up more than one resource". For example, securitygroups associated with another resource are a perfectly legit resource to look up. 

For your issue:

* It is the job of `PrefixList.fromLookup()` to handle the >=1 case.
* As for the 0 results case: we can fail the provider, that's an easy way of making sure that the 0-length array is not persisted. To achieve that, we should add an additional field into the query that indicates that a 0-element list should be considered an error.

rix0rrr commented on packages/aws-cdk/test/context-providers/cc-api-provider.test.ts:
```diff
@@ -107,10 +107,6 @@ test('looks up RDS instance using CC API listResources', async () => {
         Identifier: 'my-db-instance-2',
         Properties: '{"DBInstanceArn":"arn:aws:rds:us-east-1:123456789012:db:test-instance-2","StorageEncrypted":"false","Endpoint":{"Address":"address2.amazonaws.com","Port":"5432"}}',
       },
-      {
-        Identifier: 'my-db-instance-3',
-        Properties: '{"DBInstanceArn":"arn:aws:rds:us-east-1:123456789012:db:test-instance-3","StorageEncrypted":"true","Endpoint":{"Address":"address3.amazonaws.com","Port":"6000"}}',
-      },
```
The fact that you need to break tests (i.e., behavior) to achieve what you want to do should already tell you we can't accept this 😉

Tietew committed:
options to restrict results

Tietew committed:
Merge branch 'main' into cc-api-single-result

Tietew commented:
@rix0rrr Thank you for clarification.

I've pushed following updates:
- reverted the test I changed
- added `allowEmptyResult` and `exactResult` options to `CcApiContextQuery`
- updated tests to ensure above options

Current PRs using CC API Context Provider, except for my PrefixList, use `exactIdentifier` and is not affected by this change.
For my `PrefixList.fromLookup()`, specifying `exactResult: true` could ensure to expect 1 resources.

rix0rrr commented on packages/@aws-cdk/cloud-assembly-schema/lib/cloud-assembly/context-queries.ts:
```diff
@@ -385,6 +385,18 @@ export interface CcApiContextQuery extends ContextLookupRoleOptions {
    */
   readonly propertiesToReturn: string[];
+
+  /**
+   * Whether to allow to return empty result when propertyMatch is specified.
+   * @default false
+   */
+  readonly allowEmptyResult?: boolean;
+
+  /**
+   * Whether to restrict to return exact one result when propertyMatch is specified.
+   * @default false
+   */
+  readonly exactResult?: boolean;
```
Is this an enum in hiding?

I wonder if the following isn't more readable: 

```ts
expectedMatchCount?: 'any' | 'at-least-one' | 'at-most-one' | 'exactly-one';
```

Tietew committed:
Merge branch 'main' into cc-api-single-result

Tietew committed:
expectedMatchCount

Tietew committed:
fix docstrings

rix0rrr commented:
I'm going to wait for https://github.com/aws/aws-cdk-cli/pull/211 to be in, then refactor this PR a bit on top of that other one.

They're too uncomfortably close for me so I'm taking the reins to make them play nicely together.

Tietew commented:
Ok, thanks for adjustment!

rix0rrr committed:
Merge remote-tracking branch 'origin/main' into pr/Tietew/251

rix0rrr committed:
Merge from upstream

rix0rrr committed:
Add tests

rix0rrr committed:
Undo accidental changes

rix0rrr committed:
Merge remote-tracking branch 'origin/main' into pr/Tietew/251

rix0rrr committed:
Update build packages
